### DISCUSSION TITLE
Salesforce metadata typescript types

### DISCUSSION THREAD

**Discussion by tomcarman** [forcedotcom/cli#2937] [Q&A] [answered]
Sorry if this is not the right place to post this - but it's something I have come up against when doing some custom sf cli plugin development.

Are there any Salesforce or community efforts to create typescript types for Salesforce metadata?

I came across this - https://github.com/amphro/salesforce-to-types - but this would be used to generate something like a type for an Account object. What I am looking for, is types for the metadata itself - eg. sobject, custom field, profile etc - to be used in a sf cli plugin that interacts with metadata.

For example, a type for custom field would include things like; `fullName`, `description`, `required`, `label`, `unique` etc. (eg. https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_fielddefinition.htm or https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/customfield.htm)

If there isn't anything existing, are there any sensible ways to generate these programatically?

I've tried calling the Tooling API to describe the CustomField and FieldDefinition objects, but the fields in the response are different to the fields listed on the pages linked above.

Thanks
Tom

**Comment by mshanemc** [comment, marked as answer]
Yeah! We started that pretty recently.  Currently, `@salesforce/types` is generated from the WSDL.  A lot of that was initiated/contributed by Andrés Catalán.

we're starting to use it like
https://github.com/salesforcecli/plugin-sobject/blob/a8289d74cd0ba9c209b8b15e5dca34cc784300c4/src/shared/types.ts#L8
https://github.com/salesforcecli/plugin-sobject/blob/a8289d74cd0ba9c209b8b15e5dca34cc784300c4/src/shared/prompts/nameField.ts#L12
etc.

they're also **mostly** correct in the newer `jsforce` (v3) and (recommended for cli-plugins) `@jsforce/jsforce-node` packages but somewhat out of date.  For long-standing things that haven't changed much (ex: CustomObject) you might find it workable.

Tooling API !== metadataAPI, schema wise.  The tooling API is also `typed` for CRUD operations, so that it's hard to tell what's required or not.

One of our next big tasks in consolidating jsforce to use the `@salesforce/types` code instead of having 2 slightly different copies.

**Reply by tomcarman** [reply]
Ah this is great thank you - exactly what I was looking for.

I had actually stumbled across `@salesforce/types` but completely missed that it was covering metadata types!

Given it's directly generated from the WSDL, I assume it will get updated as new attributes are added in future Salesforce releases? Seems like using this over the types built in to jsforce is sensible - especially if your long-term plan is to consolidate.

**Reply by WillieRuemmele** [reply]
Hey @tomcarman - yes, it _should_ update as new WSDLs are published... there hasn't been a new WSDL yet to prove that's working though ;)
